extensions-contrib/oak-incremental-index/src/main/java/org/apache/druid/segment/incremental/oak/OakIncrementalIndex.java

```diff
@@ -0,0 +1,625 @@
+    final BufferAggregator[] aggregators = getAggs();
+
+    // It might be possible to rewrite this function to return a serialized row.
+    Function<Map.Entry<OakUnscopedBuffer, OakUnscopedBuffer>, Row> transformer = entry -> {
+      OakUnsafeDirectBuffer keyOakBuff = (OakUnsafeDirectBuffer) entry.getKey();
+      OakUnsafeDirectBuffer valueOakBuff = (OakUnsafeDirectBuffer) entry.getValue();
+      long serializedKeyAddress = keyOakBuff.getAddress();
+
+      long timeStamp = OakKey.getTimestamp(serializedKeyAddress);
+      int dimsLength = OakKey.getDimsLength(serializedKeyAddress);
+
+      Map<String, Object> theVals = Maps.newLinkedHashMap();
+      for (int i = 0; i < dimsLength; ++i) {
+        Object dim = OakKey.getDim(serializedKeyAddress, i);
+        DimensionDesc dimensionDesc = dimensionDescsList.get(i);
+        if (dimensionDesc == null) {
+          continue;
+        }
+        String dimensionName = dimensionDesc.getName();
+        DimensionHandler handler = dimensionDesc.getHandler();
+        if (dim == null || handler.getLengthOfEncodedKeyComponent(dim) == 0) {
+          theVals.put(dimensionName, null);
+          continue;
+        }
+        final DimensionIndexer indexer = dimensionDesc.getIndexer();
+        Object rowVals = indexer.convertUnsortedEncodedKeyComponentToActualList(dim);
+        theVals.put(dimensionName, rowVals);
+      }
+
+      ByteBuffer valueBuff = valueOakBuff.getByteBuffer();
+      int valueOffset = valueOakBuff.getOffset();
+      for (int i = 0; i < aggregators.length; ++i) {
+        Object theVal = aggregators[i].get(valueBuff, valueOffset + aggregatorOffsetInBuffer[i]);
+        theVals.put(metrics[i].getName(), theVal);
+      }
+
+      return new MapBasedRow(timeStamp, theVals);
+    };
+
+    return () -> transformIterator(descending, transformer);
+  }
+
+  // Aggregator management: initialization and aggregation
+
+  @Override
+  protected BufferAggregator[] initAggs(AggregatorFactory[] metrics,
+                                        Supplier<InputRow> rowSupplier,
+                                        boolean deserializeComplexMetrics,
+                                        boolean concurrentEventAdd)
+  {
+    this.selectors = new HashMap<>();
+    this.aggregatorOffsetInBuffer = new int[metrics.length];
+
+    int curAggOffset = 0;
+    for (int i = 0; i < metrics.length; i++) {
+      aggregatorOffsetInBuffer[i] = curAggOffset;
+      curAggOffset += metrics[i].getMaxIntermediateSizeWithNulls();
+    }
+    this.aggregatorsTotalSize = curAggOffset;
+
+    for (AggregatorFactory agg : metrics) {
+      ColumnSelectorFactory columnSelectorFactory = makeColumnSelectorFactory(
+          agg,
+          rowSupplier,
+          deserializeComplexMetrics
+      );
+
+      this.selectors.put(
+          agg.getName(),
+          new OnheapIncrementalIndex.CachingColumnSelectorFactory(columnSelectorFactory, concurrentEventAdd)
```

Comment:
I'm not sure why `CachingColumnSelectorFactory` is part of `OnheapIncrementalIndex` and not the parent class `IncrementalIndex`. I moved it to `IncrementalIndex`. I hope this will also make sense to the other reviewers.